GoogleForCreators/web-stories-wp
# Design System: Menu for DropDown

## Summary

Adds menu functionality to `<DropDown />`. 

There's a variety of dropdown versions used throughout the dashboard and the editor currently. The idea here is to be able to decrease the amount of "one off" dropdowns in play by creating something with functionality that doesn't change but appearance that can.

In the new designs we have 2 types of dropdowns: 1) that has a single list and 2) one with sublists that will need to have sub list titles displayed. In each of these there can only be one option selected at a time.

Separately we have our typeahead/search dropdown, because the search functionality adds a good amount of complexity (lists that we should virtualize, the ability to have multiple values selected) this "version" of a dropdown shouldn't be considered a version at all, but it's own thing.

Lastly, we have popover menus which are only a thing in the dashboard right now, these are essentially just the "menu" part of the dropdown, these are separate from the dropdown too. As this is still a draft, I'm weighing pros vs cons of pulling out the useDropdownMenu hook I created to be shared.

## Relevant Technical Choices

-  `<DropDownMenu />` is responsible for the menu itself. It renders a list with its items by default. It can be overridden by including a renderItem component in <Dropdown />. There's examples of this in storybook and in tests. You can also override the styles by using menuStylesOverride with styled component's CSS prop. These overrides are here because it seems in our existing ecosystem one of the leading causes of making new dropDowns is the styling will vary (IE animation effects).

- `dropDown/menu/useDropDownMenu` takes over all of the keyboard functionality. In Menu we pass the hook props and it returns to us the focusedIndex (focusedValue is also available, but I've found it unnecessary to grab yet). Inside the hook we're finding the proper index to focus in the menu and focusing it and using the listRef that belongs to the <ul> inside of Menu to handle keyboard navigation. Having this logic separate from the UI should make things easier to share and avoid needless duplication of functionality.

## To-do

- [ ] Disabled item handling 

## User-facing changes

None, changes live in storybook

## Testing Instructions

Run storybook locally and navigate to designsystem/components/dropdown. These dropdown demos should mimic our existing dropdown functionality.

- On default, if no value is currently selected the first list item should be focused
- On default, when a new item is selected the menu should close
- if isKeepMenuOpenOnSelection is true, the menu should not close when a new item is selected
- A menu should not open if the disabled knob is true in storybook
- 'No options available' should show up when menu is expanded if there are no items in an array
- If there is a currently selected value when a menu is opened it should be in focus and in view.
- You should be able to navigate dropdown list with your arrow keys. If isRTL knob is false then Up and Left arrows should move you up in the list. If isRTL knob is true then Up and Right arrows should move you up the list. (And respectfully, Down and Right arrows move you down, or Down and Left arrows move you down if isRTL is true).
- You should be able to select a new item by pressing Enter
- You should be able to close the menu by pressing Tab or Escape

---



Addresses #4950
